### ISSUE TITLE
SELECT FOR UPDATE should never be cached

### ISSUE THREAD

**Issue by jklardie** [BertrandBordage/django-cachalot#61]
The SELECT FOR UPDATE mechanism is negated by cachalot because it caches the results of queries that include SELECT FOR UPDATE. This results in unwanted concurrent behavior in systems that rely on the exclusive locking of rows. These bugs are very hard to reproduce but can and do occur.

According to postgres documentation (https://www.postgresql.org/docs/9.3/static/explicit-locking.html):

> To acquire an exclusive row-level lock on a row without actually modifying the row, select the row with SELECT FOR UPDATE. Note that once the row-level lock is acquired, the transaction can update the row multiple times without fear of conflicts.

Proper behaviour should be to let all SELECT FOR UPDATE queries through to the database. The testcase in the following link should fail (it should produce a single query, instead of using the cached result): 

https://github.com/BertrandBordage/django-cachalot/blob/7e3137eac7bc1b4f683a336579653623e71c991e/cachalot/tests/read.py#L539

A temporary workaround for now is to set the CACHALOT_QUERY_KEYGEN method, and to ignore queries containing select for updates:

```
def get_query_cache_key(compiler):
    if compiler.query.select_for_update:
        raise UncachableQuery
    return original_query_keygen(compiler)
```

**Comment by erikwt** [comment]
I can vouch for this. We ran into concurrency issues on our production environment. Not caching select_for_update like described by OP fixed this.

**Comment by BertrandBordage** [comment]
@jklardie You’re right, thanks for the detailed report. I assumed no-one would execute two times the same `SELECT FOR UPDATE` in the same transaction, but it was a bad assumption. I’ll soon release a new version with this fixed!

**Comment by jklardie** [comment]
FYI: in my system this occurs where such queries are triggered via Ajax requests. In environments where the user has bad connection, multiple requests sometimes get stacked and reach our server simultaneously.

**Comment by BertrandBordage** [comment]
@jklardie: Why did you add `select_for_update_nowait`? From what I understand, it cannot be `True` without `select_for_update` (see [lines 828-840 of query.py](https://github.com/django/django/blob/a5e13a0b926156f7885a0f1d8b300221f74601fc/django/db/models/query.py#L828-L840)).

**Comment by jklardie** [comment]
You are right, I was not aware of this. I wrongly assumed Django handled them exclusively. I will update my initial post.

**Comment by BertrandBordage** [comment]
Fixed by c5d5ee7907ae3c6304419fc6f4025eb1b0be8de8.

**Comment by BertrandBordage** [comment]
Version 1.4.0 is out, including this fix!
